integration/v3_grpc_test.go

```diff
@@ -193,3 +195,135 @@ func TestV3DeleteRange(t *testing.T) {
 		clus.Terminate(t)
 	}
 }
+
+// TestV3WatchFromCurrentRevision tests Watch APIs from current revision.
+func TestV3WatchFromCurrentRevision(t *testing.T) {
+	testKey, testKeyL, testKey2 := "foo", "fooLong", "helloworld"
+	testVal := "bar"
+
+	tests := []struct {
+		putKeys        []string
+		watchRequest   *pb.WatchRequest
+		headerRevision int64
+		createdN       int
+		wevents        []*storagepb.Event
+	}{
+		// watch the key, matching
+		{
+			[]string{testKey},
+			&pb.WatchRequest{CreateRequest: &pb.WatchCreateRequest{Key: []byte(testKey)}},
+			1,
+			1,
+			[]*storagepb.Event{
+				&storagepb.Event{
+					Kv: &storagepb.KeyValue{Key: []byte(testKey), Value: []byte(testVal), CreateRevision: 1, ModRevision: 1, Version: 1},
+				},
+			},
+		},
+		// watch the key, non-matching
+		{
+			[]string{testKey},
+			&pb.WatchRequest{CreateRequest: &pb.WatchCreateRequest{Key: []byte(testKey2)}},
+			1,
+			0,
+			[]*storagepb.Event{},
+		},
+		// watch the prefix, matching
+		{
+			[]string{testKeyL},
+			&pb.WatchRequest{CreateRequest: &pb.WatchCreateRequest{Prefix: []byte(testKey)}},
+			1,
+			1,
+			[]*storagepb.Event{
+				&storagepb.Event{
+					Kv: &storagepb.KeyValue{Key: []byte(testKeyL), Value: []byte(testVal), CreateRevision: 1, ModRevision: 1, Version: 1},
+				},
+			},
+		},
+		// watch the prefix, non-matching
+		{
+			[]string{testKey},
+			&pb.WatchRequest{CreateRequest: &pb.WatchCreateRequest{Prefix: []byte(testKey2)}},
+			1,
+			0,
+			[]*storagepb.Event{},
+		},
+	}
+
+	for i, tt := range tests {
+		clus := newClusterGRPC(t, &clusterConfig{size: 3})
+
+		wAPI := pb.NewWatchClient(clus.RandConn())
+		wStream, err := wAPI.Watch(context.TODO())
+		if err != nil {
+			t.Fatalf("#%d: wAPI.Watch error: %v", i, err)
+		}
+
+		if err := wStream.Send(tt.watchRequest); err != nil {
+			t.Fatalf("#%d: wStream.Send error: %v", i, err)
+		}
+
+		kvc := pb.NewKVClient(clus.RandConn())
+		for _, k := range tt.putKeys {
+			reqput := &pb.PutRequest{Key: []byte(k), Value: []byte(testVal)}
+			if _, err := kvc.Put(context.TODO(), reqput); err != nil {
+				t.Fatalf("#%d: couldn't put key (%v)", i, err)
+			}
+		}
+
+		evs := []*storagepb.Event{}
+		var createdWatchId int64
+		createdCnt := 0
+
+		cnt := 0
+		for cnt != len(tt.wevents) {
+			resp, err := wStream.Recv()
+			if err == io.EOF {
+				t.Fatal("unexpected io.EOF")
+			}
+			if err != nil {
+				t.Errorf("#%d: wStream.Recv error: %v", i, err)
+			}
+
+			if resp.Header == nil {
+				t.Fatalf("#%d: unexpected nil resp.Header", i)
+			} else {
+				if resp.Header.Revision != tt.headerRevision {
+					t.Logf("[TODO - skip for now] #%d: resp.Header.Revision got = %d, want = %d", i, resp.Header.Revision, tt.headerRevision)
+				}
+			}
+
+			switch {
+			case resp.Created:
+				createdWatchId = resp.WatchId
```

Comment:
well. if the assumption is that only one watcher should be created, then we should remove these `cnt` stuff. If the assumption is multiple watchers might be created, then line 298 can overwrite the previous ID.